File: kmp/composeApp/src/commonMain/kotlin/dev/jianastrero/trainer/domain/enumeration/PokemonStat.kt
package dev.jianastrero.trainer.domain.enumeration

import kotlinx.serialization.ExperimentalSerializationApi
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.JsonNames

@OptIn(ExperimentalSerializationApi::class)
@Serializable
enum class PokemonStat(val label: String) {
    @JsonNames("hp", "HP") HP("HP"),
    @JsonNames("attack", "Attack") Attack("Attack"),
    @JsonNames("defense", "Defense") Defense("Defense"),
    @JsonNames("special-attack", "SpecialAttack") SpecialAttack("Special Attack"),
    @JsonNames("special-defense", "SpecialDefense") SpecialDefense("Special Defense"),
    @JsonNames("speed", "Speed") Speed("Speed");

    companion object {
        val labels = arrayOf(
            HP,
            Attack,
            Defense,
            SpecialAttack,
            SpecialDefense,
            Speed
        )
    }
}

File: react_native/src/domain/model/enumeration/PokemonStat.ts
export class PokemonStat {

    public label: string;

    protected constructor(label: string) {
        this.label = label;
    }

    public static HP: PokemonStat = new PokemonStat("HP");
    public static Attack: PokemonStat = new PokemonStat("Attack");
    public static Defense: PokemonStat = new PokemonStat("Defense");
    public static SpecialAttack: PokemonStat = new PokemonStat("Special Attack");
    public static SpecialDefense: PokemonStat = new PokemonStat("Special Defense");
    public static Speed: PokemonStat = new PokemonStat("Speed");

    public static all(): PokemonStat[] {
        return [
            PokemonStat.HP,
            PokemonStat.Attack,
            PokemonStat.Defense,
            PokemonStat.SpecialAttack,
            PokemonStat.SpecialDefense,
            PokemonStat.Speed,
        ];
    }

    public static getFromString(stat: string): PokemonStat | undefined {
        switch (stat.toLowerCase()) {
            case 'hp':
                return PokemonStat.HP;
            case 'attack':
                return PokemonStat.Attack;
            case 'defense':
                return PokemonStat.Defense;
            case 'special-attack':
            case 'specialattack':
                return PokemonStat.SpecialAttack;
            case 'special-defense':
            case 'specialdefense':
                return PokemonStat.SpecialDefense;
            case 'speed':
                return PokemonStat.Speed;
            default:
                return undefined;
        }
    }
}

export class PokemonStatItem {
    public stat: PokemonStat;
    public baseStat: number;

    constructor(stat: PokemonStat, baseStat: number) {
        this.stat = stat;
        this.baseStat = baseStat;
    }
}
